plugins/Pandora_LayeredCalorimeterPlugin.cpp

```diff
@@ -0,0 +1,166 @@
+namespace {
+
+  static long addLayeredCalorimeterData(dd4hep::Detector& description, int argc, char** argv) {
+    const std::string LOG_SOURCE("LayeredCalorimeterPlugin");
+
+    // Ensure there are at least 4 arguments
+    if (argc < 7) {
+      printout(PrintLevel::ERROR, LOG_SOURCE, "Insufficient arguments. Expected: path, Rmin, Rmax, half_length,angle, and layer heights.");
+      return 1;
+    }
+
+    // Parse basic parameters
+    std::string subDetectorName = argv[0];
+    std::string type_flags = argv[1];
+    double Rmin = dd4hep::_toDouble(argv[2]);
+    double Rmax = dd4hep::_toDouble(argv[3]);
+    double half_length = dd4hep::_toDouble(argv[4]);
+    double angle = dd4hep::_toDouble(argv[5]);
+    std::string name_of_the_material = argv[6];
+
+    // Parse layer heights (all remaining arguments)
+    std::vector<double> layerHeights;
+    for (int i = 7; i < argc; ++i) {
+      layerHeights.push_back(dd4hep::_toDouble(argv[i]));
+    }
+
+    dd4hep::printout(PrintLevel::DEBUG, LOG_SOURCE, "Parsed: path=%s, Rmin=%f, Rmax=%f, dR=%f, angle=%f,layers=%zu",
+                    subDetectorName.c_str(), Rmin, Rmax, half_length, angle,layerHeights.size());
+
+    // Retrieve the detector element
+    auto caloDetElem = description.detector(subDetectorName);
+    if (!caloDetElem.isValid()) {
+      dd4hep::printout(PrintLevel::ERROR, LOG_SOURCE, "Invalid DetElement path: %s", subDetectorName.c_str());
+      return 1;
+    }
+     
+    unsigned int typeFlag = dd4hep::_toInt(type_flags);
+    dd4hep::printout(PrintLevel::INFO, LOG_SOURCE,"+++ setDetectorTypeFlags for detector :%s set to 0x%x",
+		     subDetectorName.c_str(),
+		     typeFlag);
+
+    caloDetElem.setTypeFlag(typeFlag);
+
+
+    // Create LayeredCalorimeterData
+    auto caloData = new dd4hep::rec::LayeredCalorimeterData;
+    caloData->layoutType = dd4hep::rec::LayeredCalorimeterData::BarrelLayout;
+    caloData->extent[0] = Rmin;
+    caloData->extent[1] = Rmax;
+    caloData->extent[2] = 0.0; // Barrel-specific value
+    caloData->extent[3] = half_length;
+
+    caloDetElem.addExtension<dd4hep::rec::LayeredCalorimeterData>(caloData);
+
+    // Material manager
+    rec::MaterialManager matMgr(caloDetElem.volume());
+    double rad_first = Rmin;
+    double rad_last = 0;
+    double dR =  Rmax - Rmin;
+    double scale_fact = dR / (-Rmin * cos(angle) + sqrt(pow(Rmax, 2) - pow(Rmin * sin(angle), 2)));
+
+    // Loop through provided layer heights
+    for (double layerHeight : layerHeights) {
+      rad_last = rad_first + (layerHeight * scale_fact);
+      dd4hep::rec::Vector3D ivr1 = dd4hep::rec::Vector3D(0., rad_first, 0); // defining starting vector points of the given layer
+      dd4hep::rec::Vector3D ivr2 = dd4hep::rec::Vector3D(0., rad_last, 0);  // defining end vector points of the given layer
+     
+      const dd4hep::rec::MaterialVec &materials = matMgr.materialsBetween(ivr1, ivr2); // calling material manager(MM) to get material info between two points
```

Comment:
do you think it would be good idea to protect against empty vectors?
```
// skip layer if no materials found
if( materials.empty() ) continue;
```

Reply:
I think unless `rad_first == rad_last` it is impossible for there to be no materials between two points? There must at least be the world volume. Or something is more wrong in the geometry than we can protect against here.